server/filestore.go

```diff
@@ -5295,6 +5313,45 @@ func (fs *fileStore) cancelAgeChk() {
 	}
 }
 
+// Lock must be held so that nothing else can interleave and write a
+// new message on this subject before we get the chance to write the
+// system tombstone. If the tombstone is written successfully then
+// this function returns a callback func to call scb and tcb after the
+// lock has been released.
+func (fs *fileStore) systemTombstoneIfNeeded(sm *StoreMsg, reason string) func() {
+	// If the deleted message was itself a tombstone then don't
+	// write out more tombstones or we'll churn endlessly.
+	if len(getHeader(JSAppliedLimit, sm.hdr)) != 0 {
+		return nil
+	}
+	if !fs.cfg.AllowMsgTTL || fs.cfg.LimitsTTL <= 0 {
+		return nil
+	}
+	if _, ok := fs.psim.Find(stringToBytes(sm.subj)); ok {
+		// There are still messages left with this subject,
+		// therefore it wasn't the last message deleted.
+		return nil
+	}
+	// Build the system tombstone.
+	var _hdr [128]byte
+	hdr := fmt.Appendf(_hdr[:0], "NATS/1.0\r\n%s: %s\r\n%s: %s\r\n\r\n", JSAppliedLimit, reason, JSMessageTTL, fs.cfg.LimitsTTL)
+	seq, ts, ttl := fs.state.LastSeq+1, time.Now().UnixNano(), int64(fs.cfg.LimitsTTL.Seconds())
+	// Store it in the stream and then prepare the callbacks
+	// to return to the caller.
+	if err := fs.storeRawMsg(sm.subj, hdr, nil, seq, ts, ttl); err != nil {
+		return nil
+	}
+	cb, tcb := fs.scb, fs.tcb
+	return func() {
```

Comment:
Why are we returning a func here vs just calling inline?

Reply:
The filestore generally seems to avoid holding the FS lock while calling the storage update handler, I assumed there was a reason for it, so I've just tried to recreate that pattern by passing back a function that the caller can call when it has released the lock.

If we don't care about holding the lock while the callbacks are called then I can remove this and make it inline.

Reply:
ok got it..